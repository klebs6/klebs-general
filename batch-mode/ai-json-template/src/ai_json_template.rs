// ---------------- [ File: ai-json-template/src/ai_json_template.rs ]
crate::ix!();

/// The derived code implements `AiJsonTemplate` for each struct, letting you
/// call `MyStruct::to_template()` to get a JSON “schema” describing how the
/// AI should produce data that matches this layout.
///
pub trait AiJsonTemplate
: Clone 
+ Debug 
+ LoadFromFile 
+ SaveToFile<Error=SaveLoadError> 
+ Serialize 
+ for<'a> Deserialize<'a> 
{
    /// Return a JSON template describing how the AI’s output should be structured.
    /// This might include doc comments or other instructions for each field.
    fn to_template() -> serde_json::Value;
}

/// Extended trait that demands both a normal `AiJsonTemplate` and
/// a `...Justification` structure for *nested* usage.
pub trait AiJsonTemplateWithJustification: AiJsonTemplate {
    fn to_template_with_justification() -> serde_json::Value;
}
