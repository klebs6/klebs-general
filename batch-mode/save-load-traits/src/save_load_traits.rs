// ---------------- [ File: src/save_load_traits.rs ]
crate::ix!();

#[async_trait]
pub trait SaveToFile {

    type Error;

    async fn save_to_file(
        &self,
        filename: impl AsRef<Path> + Send,

    ) -> Result<(), Self::Error>;
}

#[async_trait]
pub trait LoadFromFile: Sized {

    type Error;

    async fn load_from_file(filename: impl AsRef<Path> + Send) 
        -> Result<Self, Self::Error>;

}
