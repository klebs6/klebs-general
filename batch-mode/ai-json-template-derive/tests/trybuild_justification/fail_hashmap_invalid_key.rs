// ---------------- [ File: ai-json-template-derive/tests/trybuild_justification/fail_hashmap_invalid_key.rs ]
#![allow(dead_code)]
#![allow(unused_imports)]

use ai_json_template::*;
use ai_json_template_derive::*;
use std::collections::HashMap;
use serde::{Serialize, Deserialize};
use getset::{Getters, Setters};
use derive_builder::Builder;
use save_load_derive::*;
use save_load_traits::*;
use tracing::*;

/// Attempting a HashMap<bool, String>, which your macro presumably rejects as an invalid key type.
#[derive(
    SaveLoad,
    Debug,
    Clone,
    Serialize,
    Deserialize,
    AiJsonTemplate,
    AiJsonTemplateWithJustification,
    Eq,
    PartialEq,
)]
struct BadMapKeyStruct {
    kv: HashMap<bool, String>,
}

fn main() {
    println!("Should not compile: HashMap<bool, String> is invalid key type");
}
