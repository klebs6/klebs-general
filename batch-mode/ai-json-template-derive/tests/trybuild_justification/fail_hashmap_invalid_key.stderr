error: Unsupported key type in HashMap<bool, _> for AiJsonTemplate
  --> tests/trybuild_justification/fail_hashmap_invalid_key.rs:25:17
   |
25 |     kv: HashMap<bool, String>,
   |                 ^^^^

error: Unsupported key type in HashMap<bool, _> for AiJsonTemplateWithJustification
  --> tests/trybuild_justification/fail_hashmap_invalid_key.rs:22:5
   |
22 |     AiJsonTemplateWithJustification,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `AiJsonTemplateWithJustification` (in Nightly builds, run with -Z macro-backtrace for more info)

error: Unsupported key type in HashMap<bool, _> for AiJsonTemplateWithJustification
  --> tests/trybuild_justification/fail_hashmap_invalid_key.rs:25:17
   |
25 |     kv: HashMap<bool, String>,
   |                 ^^^^

error[E0277]: the trait bound `BadMapKeyStruct: Default` is not satisfied
  --> tests/trybuild_justification/fail_hashmap_invalid_key.rs:22:5
   |
22 |     AiJsonTemplateWithJustification,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `BadMapKeyStruct`
   |
help: consider annotating `BadMapKeyStruct` with `#[derive(Default)]`
   |
24 + #[derive(Default)]
25 | struct BadMapKeyStruct {
   |

error[E0369]: binary operation `==` cannot be applied to type `BadMapKeyStruct`
  --> tests/trybuild_justification/fail_hashmap_invalid_key.rs:22:5
   |
22 |     AiJsonTemplateWithJustification,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: an implementation of `PartialEq` might be missing for `BadMapKeyStruct`
  --> tests/trybuild_justification/fail_hashmap_invalid_key.rs:24:1
   |
24 | struct BadMapKeyStruct {
   | ^^^^^^^^^^^^^^^^^^^^^^ must implement `PartialEq`
help: consider annotating `BadMapKeyStruct` with `#[derive(PartialEq)]`
   |
24 + #[derive(PartialEq)]
25 | struct BadMapKeyStruct {
   |
